File: OCaml_Implementation/StatsCalculator.ml
(* ---------- Statistics Calculator in OCaml (Functional) ---------- *)
open Printf

(* ---------- Helper Functions ---------- *)

(* Read a list of floats manually *)
let rec read_list n acc =
  if n <= 0 then List.rev acc
  else (
    printf "Enter number %d: " ((List.length acc) + 1);
    let num = read_float () in
    read_list (n - 1) (num :: acc)
  )

(* Generate random float list between 1 and 100 *)
let rec generate_random_list n =
  if n <= 0 then []
  else (Random.float 100.0) :: generate_random_list (n - 1)

(* Mean: sum / length *)
let mean lst =
  let total = List.fold_left ( +. ) 0.0 lst in
  total /. float_of_int (List.length lst)

(* Median: middle value or average of two middle values *)
let median lst =
  let sorted = List.sort compare lst in
  let len = List.length sorted in
  if len mod 2 = 1 then
    List.nth sorted (len / 2)
  else
    let mid1 = List.nth sorted ((len / 2) - 1) in
    let mid2 = List.nth sorted (len / 2) in
    (mid1 +. mid2) /. 2.0

(* Mode: most frequent element(s) *)
let mode lst =
  let sorted = List.sort compare lst in
  let rec count_freq l current count acc max_count =
    match l with
    | [] ->
        if count > max_count then [current]
        else if count = max_count then current :: acc
        else acc
    | x :: xs ->
        if x = current then count_freq xs current (count + 1) acc max_count
        else if count > max_count then count_freq xs x 1 [current] count
        else if count = max_count then count_freq xs x 1 (current :: acc) max_count
        else count_freq xs x 1 acc max_count
  in
  match sorted with
  | [] -> []
  | x :: xs -> count_freq xs x 1 [] 1

(* ---------- Main Program ---------- *)

let () =
  Random.self_init ();
  printf "Enter the number of elements: ";
  let n = read_int () in
  printf "Choose input method:\n1. Manual entry\n2. Random generation\n> ";
  let choice = read_int () in
  let numbers =
    if choice = 1 then read_list n []
    else generate_random_list n
  in

  printf "\nOriginal Numbers: ";
  List.iter (fun x -> printf "%.2f " x) numbers;
  printf "\n";

  (* Sort once for display and median/mode calculations *)
  let sorted_numbers = List.sort compare numbers in
  printf "\n";
  printf "Sorted Numbers: ";
  List.iter (fun x -> printf "%.2f " x) sorted_numbers;
  printf "\n";

  let m = mean sorted_numbers in
  let med = median sorted_numbers in
  let modes = mode sorted_numbers in

  printf "\nMean: %.2f\n" m;
  printf "Median: %.2f\n" med;
  printf "Mode(s): ";
  List.iter (fun x -> printf "%.2f " x) modes;
  printf "\n"

File: Python_Implementation/StatsCalculator.py
import random

class StatisticsCalculator:
    def __init__(self, numbers):
        """Initialize with a list of numbers"""
        self.numbers = numbers

    def mean(self):
        """Calculate the mean (average)"""
        return sum(self.numbers) / len(self.numbers) if self.numbers else 0

    def median(self):
        """Calculate the median"""
        sorted_nums = sorted(self.numbers)
        n = len(sorted_nums)
        if n == 0:
            return None
        mid = n // 2
        if n % 2 == 0:
            return (sorted_nums[mid - 1] + sorted_nums[mid]) / 2
        else:
            return sorted_nums[mid]

    def mode(self):
        """Calculate the mode(s) using a dictionary"""
        if not self.numbers:
            return []
        freq = {}
        for num in self.numbers:
            freq[num] = freq.get(num, 0) + 1
        max_count = max(freq.values())
        return [num for num, count in freq.items() if count == max_count]

    def display_results(self):
        """Display the statistics"""
        sorted_nums = sorted(self.numbers)
        print("\n--- Statistics Results ---")
        print(f"Original Numbers: {self.numbers}")
        print(f"Sorted Numbers: {sorted_nums}")
        print(f"Mean: {self.mean():.2f}")
        print(f"Median: {self.median()}")
        print(f"Mode: {self.mode()}\n")


def main():
    print("=== Statistics Calculator (Python OOP) ===")
    size = int(input("Enter the size of the list: "))

    print("Choose an option:")
    print("1. Enter numbers manually")
    print("2. Generate random numbers")
    choice = input("Enter choice (1 or 2): ")

    if choice == '1':
        numbers = []
        for i in range(size):
            num = float(input(f"Enter number {i + 1}: "))
            numbers.append(num)
    elif choice == '2':
        lower = int(input("Enter the lower bound for random numbers: "))
        upper = int(input("Enter the upper bound for random numbers: "))
        numbers = [random.randint(lower, upper) for _ in range(size)]
        print(f"Generated list: {numbers}")
    else:
        print("Invalid choice. Exiting.")
        return

    calc = StatisticsCalculator(numbers)
    calc.display_results()


if __name__ == "__main__":
    main()
